staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -762,12 +762,27 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
 	// Watch cache only supports ResourceVersionMatchNotOlderThan (default).
 	unsupportedMatch := match != "" && match != metav1.ResourceVersionMatchNotOlderThan
 
-	return consistentReadFromStorage || hasContinuation || hasLimit || unsupportedMatch
+	return consistentReadFromStorage || hasContinuation || unsupportedMatch
+}
+
+// shouldApplyListLimit determines whether the cacher should
+// apply a limit to an incoming LIST request.
+//
+// note that this function doesn't check RVM nor the Continuation token.
+// these parameters are validated by the shouldDelegateList function.
+//
+// as of today, the limit is ignored for requests that set RV == 0
+func shouldApplyListLimit(opts storage.ListOptions) bool {
```

Comment:
I still think that returning int would be better:
```
if opts.Predicate.Limit <= 0 || opts.ResourceVersion == "0" {
  return 0
}
return opts.Predicate.Limit
```

and then below you effectively change to:
```
limit := computeListLimit(opts)
...
if limit > 0 && len(selectedObjects) >= limit {
   ...
}
```